File: Program_4/4(d).py
import time
import matplotlib.pyplot as plt

def fib_bottom_up(n):
    if n == 0:
        print(0, end=" ")
        return
    if n == 1:
        print(0, 1, end=" ")
        return

    dp = [0] * (n + 1)
    dp[1] = 0
    dp[2] = 1

    for i in range(3, n + 1):
        dp[i] = dp[i - 1] + dp[i - 2]

    #Print the sequence
    #for i in range(1, n + 1):
        #print(dp[i], end=" ")


sizes = [10, 100, 500, 1000]
times = []

for n in sizes:
    start = time.perf_counter()
    fib_bottom_up(n)
    end = time.perf_counter()
    elapsed = end - start
    times.append(elapsed)
    print(f"For input size {n}, time taken = {elapsed:.6f} sec")

plt.plot(sizes, times, marker='o', label="Recursive Sequence Generator")
plt.xlabel("Input Size (n)")
plt.ylabel("Time Taken (seconds)")
plt.title("Bottom-Up Fibonacci Sequence Generator Performance")
plt.legend()
plt.grid(True)
plt.show()

File: Program_4/4(d).c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void fib_bottom_up(int n) {
    if (n == 0) return;
    if (n == 1) return;

    int* dp = (int*)malloc((n + 1) * sizeof(int));
    dp[0] = 0;
    dp[1] = 1;

    for (int i = 2; i <= n; i++) {
        dp[i] = dp[i - 1] + dp[i - 2];
    }
    //printf("%d",dp[n]);
    //free(dp);
}

int main() {
    int sizes[] = {109, 1000, 5000, 10000};
    int num_sizes = 4;
    int repetitions = 10000;  // Run each input size 10000 times

    for (int i = 0; i < num_sizes; i++) {
        int n = sizes[i];

        clock_t start = clock();
        for (int j = 0; j < 1; j++) {
            fib_bottom_up(n);
        }
        clock_t end = clock();

        double time = ((double)(end - start)) / CLOCKS_PER_SEC / repetitions;
        printf("for size=%d time=%f\n",n, time);  
    }

    return 0;
}
